src/gpio.c

```diff
@@ -0,0 +1,196 @@
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
+ * WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+#include <stdio.h>
+#include <string.h>
+#include <fcntl.h>
+#include <poll.h>
+#include <inttypes.h>
+#include "gpio.h"
+
+#define PATH_MAX    64
+#define BUFFER_MAX  8
+
+int8_t gpio_export(uint8_t pin)
+{
+    int fd;
+    int8_t bytes_written, status = 0;
+    uint8_t buffer[BUFFER_MAX];
+
+    fd = open("/sys/class/gpio/export", O_WRONLY);
+    if (fd == -1)
+    {
+        fprintf(stderr, "Failed to open export for writing\n");
+        return -1;
+    }
+
+    bytes_written = snprintf(buffer, BUFFER_MAX, "%" PRIu8, pin);
+    status = write(fd, buffer, bytes_written);
+    if (status == -1)
+    {
+        fprintf(stderr, "Failed to export gpio\n");
+    }
+
+    close(fd);
+    return status;
+}
+
+int8_t gpio_unexport(uint8_t pin)
+{
+    int fd;
+    int8_t bytes_written, status = 0;
+    uint8_t buffer[BUFFER_MAX];
+
+    fd = open("/sys/class/gpio/unexport", O_WRONLY);
+    if (fd == -1)
+    {
+        fprintf(stderr, "Failed to open unexport for writing");
+        return -1;
+    }
+
+    bytes_written = snprintf(buffer, BUFFER_MAX, "%" PRIu8, pin);
+    status = write(fd, buffer, bytes_written);
+    if (status == -1)
+    {
+        fprintf(stderr, "Failed to unexport gpio");
+    }
+
+    close(fd);
+    return status;
+}
+
+int8_t gpio_direction(uint8_t pin, uint8_t direction)
+{
+    int fd;
+    int8_t status = 0;
+    char* directions_str[] = {"in", "out"};
+    char path[PATH_MAX];
+
+    snprintf(path, PATH_MAX, "/sys/class/gpio/gpio%" PRIu8 "/direction", pin);
+    fd = open(path, O_WRONLY);
+    if (fd == -1)
+    {
+        fprintf(stderr, "Failed to open gpio direction for writing\n");
+        return -1;
+    }
+
+    status = write(fd, directions_str[direction], strlen(directions_str[direction]));
+    if (status == -1)
+    {
+        fprintf(stderr, "Failed to set gpio direction\n");
+    }
+
+    close(fd);
+    return status;
+}
+
+int8_t gpio_edge(uint8_t pin, uint8_t type)
+{
+    int fd;
+    int8_t status = 0;
+    char* type_str[] = {"rising", "falling"};
+    char path[PATH_MAX];
+
+    snprintf(path, PATH_MAX, "/sys/class/gpio/gpio%" PRIu8 "/edge", pin);
+    fd = open(path, O_WRONLY);
+    if (fd == -1)
+    {
+        fprintf(stderr, "Failed to open gpio edge for writing\n");
+        return -1;
+    }
+
+    status = write(fd, type_str[type], strlen(type_str[type]));
+    if (status == -1)
+    {
+        fprintf(stderr, "Failed to set gpio edge type\n");
+    }
+
+    close(fd);
+    return status;
+}
+
+int8_t wait_for_interrupt(uint8_t gpio, uint8_t timeout)
```

Comment:
the whole gpio.c file looks like a set of lib methods that you'd use in other apps, so why do you have motion detector logic embedded in here instead of having a registered client callback?